packages/babel-plugin-transform-react-constant-elements/src/index.js

```diff
@@ -11,15 +11,39 @@ export default function () {
         return;
       }
 
+      // Elements with refs are not safe to hoist.
       if (path.isJSXIdentifier({ name: "ref" }) && path.parentPath.isJSXAttribute({ name: path.node })) {
         return stop();
       }
 
+      // Ignore identifiers & JSX expressions.
       if (path.isJSXIdentifier() || path.isIdentifier() || path.isJSXMemberExpression()) {
         return;
       }
 
-      if (!path.isImmutable()) stop();
+      if (!path.isImmutable()) {
+        // If it's not immutable, it may still be a pure expression.
+        if (path.isPure()) {
+          // If it's statically evaluatable, then evaluate it.
+          const expressionResult = path.evaluate();
+          if (expressionResult.confident) {
+            const resultNode = t.valueToNode(expressionResult.value);
+            // If it evaluated to an immutable value, we can use it.
+            // If not, it is not safe to replace as mutable values could
+            // be mutated after render.
+            // https://github.com/facebook/react/issues/3226
+            if (t.isImmutable(resultNode)) {
+              path.replaceWith(resultNode);
```

Comment:
Agreed - this is something for Babili or another uglifier. Have pushed a change.